File: app/services/sync_config.rb
# Reads and writes the sync configuration from a JSON file.
#
# Stored outside the database (storage/sync_config.json) so the sync
# folder setting survives database replacements during sync.
#
# Usage:
#   SyncConfig.sync_folder              # => "/path/to/folder" or nil
#   SyncConfig.sync_folder = "/path"    # persists to disk
#   SyncConfig.enabled?                 # => true/false
#
class SyncConfig
  CONFIG_PATH = Rails.root.join('storage', 'sync_config.json').freeze

  class << self
    def sync_folder
      read_config['sync_folder']
    end

    def sync_folder=(path)
      config = read_config
      config['sync_folder'] = path.presence
      write_config(config)
    end

    def enabled?
      folder = sync_folder
      folder.present? && Dir.exist?(folder)
    end

    def last_synced_at
      ts = read_config['last_synced_at']
      ts ? Time.parse(ts) : nil
    rescue ArgumentError
      nil
    end

    def last_synced_at=(time)
      config = read_config
      config['last_synced_at'] = time&.iso8601
      write_config(config)
    end

    private

    def read_config
      return {} unless File.exist?(CONFIG_PATH)

      JSON.parse(File.read(CONFIG_PATH))
    rescue JSON::ParserError
      {}
    end

    def write_config(config)
      FileUtils.mkdir_p(File.dirname(CONFIG_PATH))
      File.write(CONFIG_PATH, JSON.pretty_generate(config))
    end
  end
end

File: desktop/electron/services/sync-config.js
// Sync config stored in storage/sync_config.json (outside DB).

const fs = require('fs')
const path = require('path')
const db = require('../db')

function configPath() {
  return path.join(db.getStoragePath(), 'sync_config.json')
}

function readConfig() {
  try {
    if (!fs.existsSync(configPath())) return {}
    return JSON.parse(fs.readFileSync(configPath(), 'utf-8'))
  } catch {
    return {}
  }
}

function writeConfig(config) {
  const p = configPath()
  fs.mkdirSync(path.dirname(p), { recursive: true })
  fs.writeFileSync(p, JSON.stringify(config, null, 2))
}

function getSyncFolder() {
  return readConfig().sync_folder || null
}

function setSyncFolder(folder) {
  const config = readConfig()
  config.sync_folder = folder || null
  writeConfig(config)
}

function isEnabled() {
  const folder = getSyncFolder()
  return !!folder && fs.existsSync(folder)
}

function getLastSyncedAt() {
  const ts = readConfig().last_synced_at
  return ts || null
}

function setLastSyncedAt(time) {
  const config = readConfig()
  config.last_synced_at = time || new Date().toISOString()
  writeConfig(config)
}

module.exports = { getSyncFolder, setSyncFolder, isEnabled, getLastSyncedAt, setLastSyncedAt }
